### ISSUE TITLE
False positive errors when `_str = str` and `str = bytes`

### ISSUE THREAD

**Issue by AlexWaygood** [microsoft/pyright#4856]
**Describe the bug**
This code confuses pyright:

```py
from typing_extensions import TypeAlias

_str: TypeAlias = str
str = bytes
x: _str = "foo"
y: _str = b"foo"
```

Output from pyright:
```
Expression of type "Literal['foo']" cannot be assigned to declared type "_str"
  "Literal['foo']" is incompatible with "_str"
```

**Expected behaviour**
Pyright shouldn't emit an error for the definition of `x`, but _should_ for the definition of `y` (this is mypy's behaviour).

**Comments**
This might seem like slightly odd code, and it is. However, [we use these aliases in typeshed's stubs for `pika`](https://github.com/AlexWaygood/typeshed/blob/6b485a882325829dcbb6dd1d59c5e54b0bcfece1/stubs/pika/pika/spec.pyi#L7-L9) in order to accurately represent the fact that the runtime package `pika` does, in fact, alias `bytes` to the name "str".

**Pyright version used**
Reproduced on 1.1.299.

**Comment by erictraut** [comment]
I'm not able to repro the problem as reported. Here's what I see when I use pyright 1.1.299 (or pyright 1.1.300) with default settings.

![image](https://user-images.githubusercontent.com/7040122/228327409-c85c9695-ed4c-4892-95f0-b8edfe6c31b3.png)

Can you confirm that the above code exhibits the problem for you?

**Comment by AlexWaygood** [comment]
Here's what I see in VSCode:

![image](https://user-images.githubusercontent.com/66076021/228328400-fc4f122c-072a-4595-a056-47daa47bb48a.png)

I don't know how to confirm which version of pyright is included with VSCode. However, this is a minimal repro of an issue I originally encountered in https://github.com/python/typeshed/pull/9957, where we were running pyright 1.1.299 in CI: https://github.com/python/typeshed/actions/runs/4534849429/jobs/7989553568.

**Comment by erictraut** [comment]
Ah, you're using this in a ".pyi" file, not a ".py" file. Within a ".pyi" file (or in a ".py" file where `from __future__ import annotations` is used, type alias definitions are evaluated in an order-independent manner — i.e. as though they are quoted. In this case, you're relying on order dependence for the evaluation of `str`.

I think pyright is doing the right thing here. 

Rather than use a `TypeAlias`, I recommend doing the following:
```python
from builtins import str as _str
```

**Comment by AlexWaygood** [comment]
Hmm, I suppose that makes sense, thanks.
